File: python_practice/0003_longest_substring_without_repeating_char.py
from typing import Dict

''' 3. Longest Substring Without Repeating Characters

    Given a string s, find the length of the longest substring 
    without repeating characters.

    Example 1:
    Input: s = "abcabcbb"
    Output: 3
    Explanation: The answer is "abc", with the length of 3.
'''


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        start = max_length = 0
        used_char: Dict[str, int] = {}

        for i, char in enumerate(s):
            if char in used_char and start <= used_char[char]:
                start = used_char[char] + 1
            else:
                max_length = max(max_length, i - start + 1)

            used_char[char] = i

        return max_length


s = Solution()
a = 'abcabcbb'
s.lengthOfLongestSubstring(a)

File: dart_practice/lib/0003_longest_substring_without_repeating_char.dart
import 'dart:math';

int lengthOfLongestSubstring(String s) {
  int start = 0;
  int maxLength = 0;
  Map<String, int> usedChar = {};

  List<String> split = s.split('');

  for (int i = 0; i < split.length; i++) {
    final char = split[i];

    if (usedChar.containsKey(char) && start <= usedChar[char]!) {
      start = usedChar[char]! + 1;
    } else {
      maxLength = max(maxLength, i - start + 1);
    }

    usedChar[char] = i;
  }

  return maxLength;
}
